v2/pkg/engine/datasource/graphql_datasource/graphql_subscription_client.go

```diff
@@ -181,22 +182,38 @@ func (c *SubscriptionClient) subscribeWS(reqCtx context.Context, options GraphQL
 }
 
 // generateHandlerIDHash generates a Hash based on: URL and Headers to uniquely identify Upgrade Requests
-func (c *SubscriptionClient) generateHandlerIDHash(options GraphQLSubscriptionOptions) (uint64, error) {
+func (c *SubscriptionClient) generateHandlerIDHash(ctx *resolve.Context, options GraphQLSubscriptionOptions) (uint64, error) {
 	var (
 		err error
 	)
 	xxh := c.hashPool.Get().(*xxhash.Digest)
 	defer c.hashPool.Put(xxh)
 	xxh.Reset()
 
-	_, err = xxh.WriteString(options.URL)
-	if err != nil {
+	if _, err = xxh.WriteString(options.URL); err != nil {
 		return 0, err
 	}
-	err = options.Header.Write(xxh)
-	if err != nil {
+	if err := options.Header.Write(xxh); err != nil {
 		return 0, err
 	}
+	// The client request headers might be propagated to the outgoing
+	// request by the caller. We need to take them into account as well.
+	// However, to improve the changes of connection reuse we do skip
+	// some headers that are very likely to be different but very unlikely
+	// to cause different responses, otherwise no connections would be reused.
+	for key, values := range ctx.Request.Header {
```

Comment:
I think we should have an allow list of headers that we want to add to the hash.
E.g. in most cases I think we want to limit this to just "Authorization".
This allow list could be configurable at the resolver level, and then we expose it through the cosmo router yaml.
WDYT?

I can also imagine use cases where we want the user to be authenticated, but the origin connection should not use the Authorization header. So the behavior should be configurable and potentially opt-out.

cc @StarpTech

Reply:
I agree, it should be configurable and I like the proposal to maintain a waitlist rather than taking all headers into consideration. I'd definitely add `Authorization` by default to prevent data exposure and let the rest be use case specific.

Reply:
I've refactored it so the caller can pass a list of headers. This allows us to make it automatic in the router side, by passing the headers declared in the forwarding rules. PTAL